### ISSUE TITLE
adversarial loss could be negative?

### ISSUE THREAD

**Issue by yifita** [alexjc/neural-enhance#31]
Thanks for sharing your code!
I'm wondering, you used `softplus` and `x-softplus` for adversarial losses, what is the theoretical reason for this?
Especially, since you are feeding discriminators' output as x without softmax normalization, this could initially yield a negative loss?
For example:
in generator: `1-softplus(G(seed))` < 0 for large disc_out 
same in discriminator: `G(seed)-softplus(G(seed))` is always negative

**Comment by alexjc** [comment]
The losses should be the very same as the usual sigmoid+log combination used for GANs. It was Casper who suggested it via Twitter. See this section of Ferenc's blog post:
www.inference.vc/are-energy-based-gans-actually-energy-based/#originalganlogisticregression

I messed up the adversarial loss in the refactor though, it should have been softminus. I just fixed it ;-)  The reason for this new form is to avoid the division problems and NaN outputs that crash the training—without having to add `epsilon` hacks everywhere.

**Comment by yifita** [comment]
right.. I got the transformation from sigmoid+log to softplus/softminus...
But I don't completely agree with Ferenc's blog which minizes logD(x)+log(1-D(G(X))) in the innerloop. shouldn't it be maximising it? (minmax problem in the original paper)
As as result, the discriminator loss should be softplus(fake)-softminus(real)

**Comment by alexjc** [comment]
The code with sigmoid+log is relatively standard (using inspiration from many existing GAN implementations) and was used to trained the current models. I checked the transformation to softplus again, they are the same but help with NaN errors.

If you think it should be better the other way around, try it and let me know!

**Comment by yifita** [comment]
other GAN implementation use sigmoid cross entropy loss (sigmoid + log)
[example here](https://github.com/bamos/dcgan-completion.tensorflow/blob/master/model.py#L87)
the definition of this function in tensorflow is:
z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x)) [tf api](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#sigmoid_cross_entropy_with_logits)

In case of real data, z = 1, we will get -log(sigmoid(x)) which will -softminus.
As for fake data, z = 0, we will get -log(1 - sigmoid(x)) which is softplus.
So in my opinion, adding these two terms will give the discriminator's loss as softplus(fake)-softminus(real)

Or there's something I missed?

**Comment by alexjc** [comment]
Hmm, I worked from an example Theano GAN implementation and since the discriminator worked well and the results of super-resolution were good I didn't look further....

The current discriminator loss is exactly your equation but negated. If the adversarial loss is correct as-is now, then that's a problem ;-) Have you tried the change?

**Comment by yifita** [comment]
the adversarial loss would be -softplus(fake) which is also the negation of what you just corrected to: x-softminus(fake). 
No I haven't tried u.u I thought I would ask you first...
Could you please publish or email me the log of training loss (and applied loss weights) if you have one? So that I could use it as reference.

**Comment by alexjc** [comment]
Ah, good. I just checked the code again too and everything seems to be in order. (You did have me worried ;-) If there's benefit in negating both those loss equations to make them more "standard" it could be worth doing, but at least it's consistent like this.

Whether it's the best model for super-resolution and this architecture I have no idea yet :-D

**Comment by alexjc** [comment]
The current code works and has scripts to reproduce training. I'm sure it could be made better, and if you have a reason to switch the signs of both losses (for same result) just open the ticket again and let me know!

**Comment by AbbasEl-Bourji** [comment]
Hello, Please go to Play Store and update the App, in the new update
version do you can download photos, videos and documents. If you update the
App you DON’T need to rescan the QR code again!  And if you have android
version 5.0 and up then the size of the app is after the update only 9 mb.
Please give us 5 stars and a good review my friend. Have fun with the App !
